contrib/golang/filters/http/source/go/pkg/http/shim.go

```diff
@@ -105,10 +105,34 @@ func requestFinalize(r *httpRequest) {
 	r.Finalize(api.NormalFinalize)
 }
 
+func getOrCreateState(s *C.processState) *processState {
+	r := s.req
+	req := getRequest(r)
+	if req == nil {
+		req = createRequest(r)
+	}
+	if s.is_encoding == 0 {
+		if req.decodingState.processState == nil {
+			req.decodingState.processState = s
+		}
+		return &req.decodingState
+	}
+
+	// s.is_encoding == 1
+	if req.encodingState.processState == nil {
+		req.encodingState.processState = s
+	}
+	return &req.encodingState
+}
+
 func createRequest(r *C.httpRequest) *httpRequest {
 	req := &httpRequest{
 		req: r,
 	}
+	req.decodingState.request = req
+	req.encodingState.request = req
+	req.streamInfo.request = req
+
```

Comment:
The cyclic refer happens inside of `httpRequest`, the GC finalizer works as expected.

```
type httpRequest struct {
	decodingState processState
	encodingState processState
	streamInfo    streamInfo
}
```

Reply:
@doujiang24 
There is a call `runtime.SetFinalizer(req, requestFinalize)` a few lines below.
In the demo, the finalizer added to `r` is not invoked. The `req` is the `r` in the demo, the `decodingState` is the `s` in the demo. So maybe the  `requestFinalize` is not invoked just like the finalizer added to `r`?

Reply:
@spacewander After changing `s *state` to `s state`, as the following demo, finalizer works as expected.

```
type state struct {
	r *req
}
type req struct {
	s state
}

func main() {
	for i := 0; i < 100; i++ {
		r := &req{}
		r.s.r = r
		runtime.SetFinalizer(r, func(r *req) {
			println("finalizer")
		})
	}
	runtime.GC()
}
```

Reply:
I see. The decodingState here is not a pointer so it's not another GC object. Before I can approve it, would you add a comment on the definition of the states so that people won't get hurt by changing the type of the states later?

Reply:
@spacewander done, PTAL again, thanks~